sources/packages/backend/libs/integrations/src/services/disbursement-schedule/e-cert-generation.service.ts

```diff
@@ -281,27 +306,109 @@ export class ECertGenerationService {
+          } as Restriction;
         } else {
           const effectiveValue =
             disbursementValue.valueAmount -
             (disbursementValue.disbursedAmountSubtracted ?? 0) -
             (disbursementValue.overawardAmountSubtracted ?? 0);
           disbursementValue.effectiveAmount = round(effectiveValue);
+
+          const application = disbursement.studentAssessment.application;
+          await this.checkLifeTimeMaximumAndAddStudentRestriction(
+            disbursementValue,
+            application,
+            entityManager,
+          );
         }
       }
     }
   }
 
+  /**
+   * Check if the student hits/exceeds the life time maximum for the BCSL Award.
+   * If they hits/exceeds the life time maximum, then the award is reduced so the
+   * student hits the exact maximum value and the student restriction
+   * {@link  RestrictionCode.BCLM} is placed for the student.
+   * @param disbursementValue disbursement value.
+   * @param application application related to the disbursement.
+   * @param entityManager used to execute the commands in the same transaction.
+   */
+  private async checkLifeTimeMaximumAndAddStudentRestriction(
+    disbursementValue: DisbursementValue,
+    application: Application,
+    entityManager: EntityManager,
+  ): Promise<void> {
+    if (disbursementValue.valueType === DisbursementValueType.BCLoan) {
+      const student = application.student;
+      const maxLifetimeBCLoanAmount =
+        application.programYear.maxLifetimeBCLoanAmount;
+      const auditUser = await this.systemUsersService.systemUser();
+
+      const totalLifeTimeAmount =
+        (await this.sfasApplicationService.totalLegacyBCSLAmount(
+          student.sinValidation.sin,
+          student.birthDate,
+          student.user.lastName,
+        )) +
+        (await this.disbursementScheduleSharedService.totalDisbursedBCSLAmount(
+          student.id,
+        )) +
+        disbursementValue.effectiveAmount;
+
+      if (totalLifeTimeAmount >= maxLifetimeBCLoanAmount) {
+        // Amount subtracted when lifetime maximum is reached.
+        const amountSubtracted = totalLifeTimeAmount - maxLifetimeBCLoanAmount;
+        // Ideally disbursementValue.effectiveAmount should be greater or equal to amountSubtracted.
+        // The flow will not reach here if the ministry ignore the restriction for the previous
+        // disbursement/application and money went out to the student, even though they reach the maximum.
+        const newEffectiveAmount =
+          disbursementValue.effectiveAmount - amountSubtracted;
+        /**
+         * Create {@link RestrictionCode.BCLM} restriction when lifetime maximum is reached/exceeded.
+         */
+        const bclmRestriction =
+          await this.studentRestrictionSharedService.createRestrictionToSave(
+            student.id,
+            RestrictionCode.BCLM,
+            auditUser.id,
+            application.id,
+          );
+
+        if (bclmRestriction) {
```

Comment:
Edge case but a possible scenario.
Depending on certain circumstances we may be processing at the same time 2 disbursements for the same students, for instance, assuming that we have two applications as below.

**Application A**:  Jan, 1st - April, 30th
**Application B**: May, 1st - July, 31st.
Wonder that Application A can be reassessed near the offering end date on April, 30th and it may be having the e-Cert generated on May, 1st, which will lead both to be part of the same e-Cert file.

The e-Cert SQL query will get both records where no BCSL restriction will be in place, but then the first one will be adding the BCSL restriction and the second one must be aware of it to stop the BC funding.

When we have 2 records for the same student, we may need to have a way to refresh the restrictions in a way that the second disbursement will not have the BC fundings disbursed.

Reply:
mm, That is a good edge case, are we managing a similar scenario in overwards.? ya we need to find a way

Reply:
@dheepak-aot  @guru-aot @andrepestana-aot  @sh16011993  I and @andrewsignori-aot were discussing this and for the PR I am planning to add a directory that will be updated when a student gets a restriction for the first time, and as this directory as the reference the next disbursement of the same student will forwarded to execute restriction logic

Reply:
As discussed we will handle this in a separate ticket

Reply:
Lets handle it in a separate ticket, can we add a TODO here?